kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt

```diff
@@ -3,86 +3,163 @@ package io.kotest.assertions.eq
-      return actual.all { elementInActualSet ->
+   private fun equalsIgnoringOrder(actual: Set<*>, expected: Set<*>, strictNumberEq: Boolean): Pair<Boolean, Throwable?> {
+
+      var innerError: Throwable? = null
+
+      fun equalWithDetection(elementInActualSet: Any?, it: Any?) =
+         eq(elementInActualSet, it, strictNumberEq)?.let {
+            if (null == innerError && (it.message?.startsWith(trigger) == true)) innerError = it
+            false
+         } ?: true
+
+      return Pair(actual.all { elementInActualSet ->
          // if we have a collection type we must use the eq typeclass
          // to ensure we can support deep equals, otherwise we can just compare
          when (elementInActualSet) {
-            is Iterable<*> -> expected.any { eq(elementInActualSet, it, strictNumberEq) == null }
-            is Array<*> -> expected.any { eq(elementInActualSet, it, strictNumberEq) == null }
+            is Set<*> -> expected.any {
+               it is Set<*> && equalWithDetection(elementInActualSet, it)
+            }
+            is Map<*,*> -> expected.any {
+               it is Map<*,*> && equalWithDetection(elementInActualSet, it)
+            }
+            is Collection<*>, is Array<*> -> expected.any {
+               it !is Set<*>
+                  && (it is Collection<*> || it is Array<*>)
+                  && equalWithDetection(elementInActualSet, it)
+            }
+            is Iterable<*> -> expected.any {
+               it !is Set<*>
+                  && it !is Collection<*>
+                  && it !is Array<*>
+                  && it is Iterable<*>
+                  && equalWithDetection(elementInActualSet, it)
+            }
             else -> expected.contains(elementInActualSet)
          }
+      }, innerError)
+   }
+
+   const val trigger = "Disallowed"
+
+   private fun errorWithTypeDetails(actual: Iterable<*>, expected: Iterable<*>): Throwable {
+      val tag = "${actual::class.simpleName?.let {it} ?: actual::class} with ${expected::class.simpleName?.let {it} ?: expected::class}\n"
+      val detailErrorMessage = when {
+         actual is Set<*> || expected is Set<*> -> "$trigger: Set can be compared only to Set\nMay not compare $tag"
+         (actual is Collection || actual is Array<*>) || (expected is Collection || expected is Array<*>) -> "$trigger typed contract\nMay not compare $tag"
+         else -> "$trigger promiscuous iterators\nMay not compare $tag"
       }
+      return failure(Expected(Printed("*")), Actual(Printed("*")), detailErrorMessage)
    }
 
+   private const val disallowed = "$trigger nesting iterator"
+
    private fun checkEquality(actual: Iterable<*>, expected: Iterable<*>, strictNumberEq: Boolean): Throwable? {
-      var index = 0
+
       val iter1 = actual.iterator()
       val iter2 = expected.iterator()
       val elementDifferAtIndex = mutableListOf<Int>()
+
+      fun <T> nestedIterator(item: T): String? = item?.let {
```

Comment:
`(it::class.isInstance(actual) || actual::class.isInstance(it)))` part seems wrong, as this method is getting used to check both actual and expected. But in method body instance checking getting perform only with `actual`
This `nestedIterator(a)` need to change to  `nestedIterator(a, actual)`
 `nestedIterator(b)` need to change to  `nestedIterator(a, expected)`

Reply:
modified as requested